controllers/space/space_controller.go

```diff
@@ -0,0 +1,343 @@
+		return false, nil // skip NSTemplateSet deletion
+	}
+	// Get the NSTemplateSet associated with the Space
+	nstmplSet := &toolchainv1alpha1.NSTemplateSet{}
+	err := memberCluster.Client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: memberCluster.OperatorNamespace,
+			Name:      space.Name},
+		nstmplSet)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return false, err // something wrong happened
+		}
+		logger.Info("the NSTemplateSet resource is already deleted")
+		return false, nil // NSTemplateSet was already deleted
+	}
+	if util.IsBeingDeleted(nstmplSet) {
+		logger.Info("the NSTemplateSet resource is already being deleted")
+		return true, nil // requeue until fully deleted
+	}
+	logger.Info("deleting the NSTemplateSet resource")
+	// Delete NSTemplateSet associated with Space
+	if err := memberCluster.Client.Delete(context.TODO(), nstmplSet); err != nil {
+		if !errors.IsNotFound(err) {
+			return false, err // something wrong happened
+		}
+		return false, nil // was already deleted in the mean time
+	}
+	logger.Info("deleted the NSTemplateSet resource")
+	return true, nil // requeue until fully deleted
+}
+
+func (r *Reconciler) setStatusProvisioning(space *toolchainv1alpha1.Space) error {
+	return r.updateStatus(
+		space,
+		toolchainv1alpha1.Condition{
+			Type:   toolchainv1alpha1.ConditionReady,
+			Status: corev1.ConditionFalse,
+			Reason: toolchainv1alpha1.SpaceProvisioningReason,
+		})
+}
+
+func (r *Reconciler) setStatusProvisioningFailed(logger logr.Logger, space *toolchainv1alpha1.Space, cause error) error {
+	if err := r.updateStatus(
+		space,
+		toolchainv1alpha1.Condition{
+			Type:    toolchainv1alpha1.ConditionReady,
+			Status:  corev1.ConditionFalse,
+			Reason:  toolchainv1alpha1.SpaceProvisioningFailedReason,
+			Message: cause.Error(),
+		}); err != nil {
+		logger.Error(cause, "unable to provision Space")
+		return err
+	}
+	return cause
+}
+
+func (r *Reconciler) setStatusReady(space *toolchainv1alpha1.Space) error {
+	return r.updateStatus(
+		space,
+		toolchainv1alpha1.Condition{
+			Type:   toolchainv1alpha1.ConditionReady,
+			Status: corev1.ConditionTrue,
+			Reason: toolchainv1alpha1.SpaceProvisionedReason,
+		})
+}
+
+func (r *Reconciler) setStatusTerminating(space *toolchainv1alpha1.Space) error {
+	return r.updateStatus(
+		space,
+		toolchainv1alpha1.Condition{
+			Type:   toolchainv1alpha1.ConditionReady,
+			Status: corev1.ConditionFalse,
+			Reason: toolchainv1alpha1.SpaceTerminatingReason,
+		})
+}
+
+func (r *Reconciler) setStatusTerminatingFailed(logger logr.Logger, space *toolchainv1alpha1.Space, cause error) error {
+	if err := r.updateStatus(
+		space,
+		toolchainv1alpha1.Condition{
+			Type:    toolchainv1alpha1.ConditionReady,
+			Status:  corev1.ConditionFalse,
+			Reason:  toolchainv1alpha1.SpaceTerminatingFailedReason,
+			Message: cause.Error(),
+		}); err != nil {
+		logger.Error(cause, "unable to terminate Space")
+		return err
```

Comment:
just out of curiosity - why did you introduce a new way of setting the failure conditions instead of using the same functions that are used in other controllers - eg.: https://github.com/codeready-toolchain/host-operator/blob/master/controllers/masteruserrecord/masteruserrecord_controller.go#L218-L226
it would be nice to keep the usage consistent.

Reply:
because each time I read the code you linked, I have to figure out what are the arguments to pass. I don't find it convenient at all.

Reply:
ok, but wouldn't it be good to keep the consistency? 
Feel free to create a story for simplification of these functions.

Reply:
in addition - in your version, you don't wrap the error if the update of the status fails.

Reply:
good idea! See https://issues.redhat.com/browse/CRT-1406

Reply:
> in addition - in your version, you don't wrap the error if the update of the status fails.

if the update fails, the error is returned and it will be logged by the controller (or manager) itself (ie, by the caller to `Reconcile`). In other words, the `cause` is logged here, and the status update error is logged afterwards. Or am I missing something?

Reply:
Yeah, you are probably right. 
It's just that I'm used to the current way of updating the status, so it's harder to compare it with your version. That's why I proposed using the same way to keep the consistency - mainly when more people will work on this code.

Reply:
I agree with you about the need to consistency but for now, can we keep this code as-is in this controller and see how we address the consistency in https://issues.redhat.com/browse/CRT-1406?

Reply:
We don't know when we will have time to work on CRT-1406 so we can stay with the mix of approaches for a few months. Just saying. :wink: :peace_symbol: 

Let's wait also for @rajivnathan's opinion on it as he will work with the code of this controller as well. I'm fine with keeping it as it is.